compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/test/iree_gpu_ops.mlir

```diff
@@ -92,3 +92,134 @@ func.func @tensor_barrier_multiple_inputs(%input: tensor<?xf16>) -> (tensor<?xf1
+func.func @coalesced_gather_dma_f16(%indices: tensor<128x64xindex>, %source: tensor<2048x64xf16>, %dest: tensor<128x64xf16>) -> tensor<128x64xf16> {
+  %c1 = arith.constant 1 : index
+  %result = scf.forall (%i) in (%c1) shared_outs(%out = %dest) -> (tensor<128x64xf16>) {
+    scf.forall.in_parallel {
+      iree_gpu.coalesced_gather_dma %indices, %source into %out
+        : tensor<128x64xindex>, tensor<2048x64xf16>, tensor<128x64xf16> -> tensor<128x64xf16>
+    }
+  }
+  return %result : tensor<128x64xf16>
+}
+
+// CHECK-LABEL: func @coalesced_gather_dma_f16
+//  CHECK-SAME:   %[[INDICES:[A-Za-z0-9]+]]: tensor<128x64xindex>
+//  CHECK-SAME:   %[[SOURCE:[A-Za-z0-9]+]]: tensor<2048x64xf16>
+//  CHECK-SAME:   %[[DEST:[A-Za-z0-9]+]]: tensor<128x64xf16>
+//       CHECK:   scf.forall
+//       CHECK:     scf.forall.in_parallel
+//       CHECK:       iree_gpu.coalesced_gather_dma %[[INDICES]], %[[SOURCE]] into %{{.+}} : tensor<128x64xindex>, tensor<2048x64xf16>, tensor<128x64xf16> -> tensor<128x64xf16>
+
+// -----
+
+// Test coalesced_gather_dma with 1D tensors
+// indices: 1024 elements, dest: 256xf32 = 256 * 4 bytes = 1024 bytes, ratio = 1024/1024 = 1
+func.func @coalesced_gather_dma_1d(%indices: tensor<1024xindex>, %source: tensor<1024xf32>, %dest: tensor<256xf32>) -> tensor<256xf32> {
+  %c1 = arith.constant 1 : index
+  %result = scf.forall (%i) in (%c1) shared_outs(%out = %dest) -> (tensor<256xf32>) {
+    scf.forall.in_parallel {
+      iree_gpu.coalesced_gather_dma %indices, %source into %out
+        : tensor<1024xindex>, tensor<1024xf32>, tensor<256xf32> -> tensor<256xf32>
+    }
+  }
+  return %result : tensor<256xf32>
+}
+
+// CHECK-LABEL: func @coalesced_gather_dma_1d
+//  CHECK-SAME:   %[[INDICES:[A-Za-z0-9]+]]: tensor<1024xindex>
+//  CHECK-SAME:   %[[SOURCE:[A-Za-z0-9]+]]: tensor<1024xf32>
+//  CHECK-SAME:   %[[DEST:[A-Za-z0-9]+]]: tensor<256xf32>
+//       CHECK:   scf.forall
+//       CHECK:     scf.forall.in_parallel
+//       CHECK:       iree_gpu.coalesced_gather_dma %[[INDICES]], %[[SOURCE]] into %{{.+}} : tensor<1024xindex>, tensor<1024xf32>, tensor<256xf32> -> tensor<256xf32>
+
+// -----
+
+// Test coalesced_gather_dma within scf.forall with tensor.parallel_insert_slice
+// Using simpler dimensions to satisfy the verifier constraint
+func.func @coalesced_gather_dma_in_forall(%indices: tensor<1024x64xindex>, %source: tensor<2048x64xf32>, %dest: tensor<256x16xf32>) -> tensor<256x16xf32> {
+  %c4 = arith.constant 4 : index
+  %result = scf.forall (%i) in (%c4) shared_outs(%out = %dest) -> (tensor<256x16xf32>) {
+    %c64 = arith.constant 64 : index
+    %c256 = arith.constant 256 : index
+    %c0 = arith.constant 0 : index
+    %c4_idx = arith.constant 4 : index
+    %offset_idx = arith.muli %i, %c256 : index
+    %offset_dest = arith.muli %i, %c4_idx : index
+
+    // Extract slices for this iteration
+    // indices: 64x16 = 1024 elements, dest: 64x16xf32 = 1024 * 4 bytes = 4096 bytes, ratio = 4096/1024 = 4
```

Comment:
Please expand this comment a bit? I don't follow why you're doing this computation.

Reply:
those are outdated, removed.